javascript/ql/src/Declarations/UnreachableMethodOverloads.ql

```diff
@@ -0,0 +1,131 @@
+/**
+ * @name Unreachable method overloads
+ * @description Having multiple overloads with the same parameter types in TypeScript
+ *              causes the latter overloads to become unreachable, as the TypeScript compiler
+ *              will always choose the textually first matching overload when selecting
+ *              which overload is called at a callsite.
+ * @kind problem
+ * @problem.severity warning
+ * @id js/unreachable-method-overloads
+ * @precision high
+ * @tags correctness
+ *       typescript
+ */
+
+import javascript
+import semmle.javascript.Variables
+
+/**
+ * Gets the i'th parameter from the method signature.
+ */
+SimpleParameter getParameter(MethodSignature sig, int i) { result = sig.getBody().getParameter(i) }
+
+/**
+ * Gets a string-representation of the type-annotation from the i'th parameter in the method signature.
+ */
+string getParameterTypeAnnotation(MethodSignature sig, int i) {
+  result = getParameter(sig, i).getTypeAnnotation().toString()
+}
+
+/**
+ * Gets the other overloads for an overloaded method signature.
+ */
+MethodSignature getOtherMatchingSignatures(MethodSignature sig) {
+  signaturesMatch(result, sig) and
+  result != sig
+}
+
+/**
+ * Gets the kind of the member-declaration. Either "static" or "instance".
+ */
+string getKind(MemberDeclaration m) {
+  if m.isStatic() then result = "static" else result = "instance"
+}
+
+/**
+ * A call-signature that originates from a MethodSignature in the AST.
+ */
+private class MethodCallSig extends CallSignatureType {
+    string name;
+    
+    MethodCallSig() {
+        exists(MethodSignature sig | 
+            this = sig.getBody().getCallSignature() and
+            name = sig.getName()
+        )
+    }
+    
+    /**
+     * Gets the name of the associated method
+     */
+    string getName() {
+        result = name
```

Comment:
In practice it won't really be relevant that `getName()` can have multiple results. 

In the query `getName()` is only used to avoid a massive cartesian product in `matchingCallSignature`. 

I've change the qldoc.